### ISSUE TITLE
`.otherwise()` does not inherit narrowed types

### ISSUE THREAD

**Issue by bdbaraban** [gvergnaud/ts-pattern#118]
**Describe the bug**
When matching a union type, narrowed types do not get filtered out of the value selected in `.otherwise()`.

**Minimal reproduction case**
https://codesandbox.io/s/ts-pattern-union-otherwise-4e5qso

```
import { match } from "ts-pattern";

type Union = "a" | "b" | "c";

const fn = (union: Union) =>
  match(union)
    .with("a", (a) => /* a: 'a' */ a)
    .with("b", (b) => /* b: 'b' */ b)
    .otherwise((c) => /* Expect `c: 'c'` but get `c: Union` */ c);
```

**Versions**
- TypeScript version: 4.8.4
- ts-pattern version: 4.0.5

**Comment by gvergnaud** [comment]
Hey, this is expected. Each branch is independent from the other and start their narrowing from the input type given to `match`.  This is a compromise I made to make type-checking faster and make type errors easier to understand.

It was unfortunately too slow to run exclude handled cases from the input type for moderately complex input types and patterns. It would be worth revisiting this assumption though since several TS versions have been released since then.
